db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/TernaryOperatorNode.java

```diff
@@ -914,38 +913,64 @@ public ValueNode substrBind()
 		** string value types.
 		*/
 		receiverType = receiver.getTypeId();
-		switch (receiverType.getJDBCTypeId())
-		{
-			case Types.CHAR:
-			case Types.VARCHAR:
-			case Types.LONGVARCHAR:
-			case Types.CLOB:
-				break;
-			default:
+		if (!receiverType.isStringTypeId() && !receiverType.isBitTypeId()) {
+			throwBadType("SUBSTR", receiverType.getSQLTypeName());
+		}
+
+
+		// Determine the maximum length of the result
+		int maximumWidth = receiver.getTypeServices().getMaximumWidth();
+		boolean isFixedLength = false;
+		int resultLen = maximumWidth;
+
+		TypeId	resultType;
+
+		// receiver is fixed type and either start or length is constant, then result should be fixed type with known length;
+		// or receiver is vartype and length is constant, then result should be fixed type with known length
+		if ((receiverType.getTypeFormatId() == StoredFormatIds.CHAR_TYPE_ID ||
+			receiverType.getTypeFormatId() == StoredFormatIds.BIT_TYPE_ID)) {
+			if (rightOperand != null && rightOperand instanceof ConstantNode)
 			{
-				throwBadType("SUBSTR", receiverType.getSQLTypeName());
+				resultLen = ((ConstantNode)rightOperand).getValue().getInt();
+				isFixedLength = true;
+			} else if (leftOperand != null && leftOperand instanceof ConstantNode) {
+				int startPostion = ((ConstantNode)leftOperand).getValue().getInt();
+				resultLen = maximumWidth - startPostion + 1;
+				isFixedLength = true;
+			}
+		} else if ((receiverType.getTypeFormatId() == StoredFormatIds.VARCHAR_TYPE_ID ||
+				receiverType.getTypeFormatId() == StoredFormatIds.VARBIT_TYPE_ID)) {
+			if (rightOperand != null && rightOperand instanceof ConstantNode)
+			{
+				resultLen = ((ConstantNode)rightOperand).getValue().getInt();
+				isFixedLength = true;
```

Comment:
If the receiver type is varbit ot varchar, should isFixedLength be false?

Reply:
If receiver is vartype and length is constant, then the result should still be fixed type with known length. This is similar to DB2 behavior.